pkg/httpserver/traffic_redistribution_middleware.go

```diff
@@ -0,0 +1,152 @@
+package httpserver
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/justtrackio/gosoline/pkg/appctx"
+	"github.com/justtrackio/gosoline/pkg/cache"
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/clock"
+	"github.com/justtrackio/gosoline/pkg/log"
+)
+
+const (
+	distributorMaxSize   int64  = 10000
+	distributorPruneSize uint32 = 1000
+)
+
+//go:generate mockery --name TrafficDistributor
+type (
+	TrafficDistributor interface {
+		// ShouldCloseConnection checks whether the connection to the remote address should be closed (after handling any currently in flight requests).
+		ShouldCloseConnection(remoteAddr string, headers http.Header) bool
+	}
+
+	noopDistributor struct{}
+
+	distributor struct {
+		clock    clock.Clock
+		settings TrafficDistributorSettings
+		tracker  cache.Cache[trafficEntry]
+	}
+
+	trafficEntry struct {
+		requestCount int
+		activeSince  time.Time
+	}
+
+	trafficDistributionKey string
+)
+
+// ProvideTrafficDistributor returns a TrafficDistributor.
+// This is a component that tracks the lifecycle of client connections
+// based on configurable policies. On that basis it provides recommendations on whether a connection
+// should be closed or not. This is useful to prevent clients from keeping connections open, eg in k8s
+// where load balancing only happens on new connections.
```

Comment:
```suggestion
// ProvideTrafficDistributor returns a TrafficDistributor.
// This is a component that tracks the lifecycle of client connections based on configurable policies. 
// On that basis it provides recommendations on whether a connection should be closed or not. 
// This is useful to prevent clients from keeping connections open for an extended period of time, 
// for example in kubernetes where load balancing only happens on new connections.
```